File: DSA/queue.cpp
#include<iostream>
using namespace std;

class Queue
{
    int Rear=0;
    int Front=0;
    int Size=5,Q[5],Y,i,New;
public:
    int Enqueue();
    void display();
    void Dequeue();
    int Search();
    int Change();
};
int Queue::Enqueue()
{
    if(Rear>=Size)
    {
        cout<<"Queue Overflow";
    }
    else
    {
        for(Rear=1;Rear<=6;Rear++)
        {
            cout<<"Enter Data "<<Rear<<"=";
            cin>>Q[Rear];
        }
    }
    if(Front==0)
    {
        Front=1;
    }

}
void Queue::Dequeue()
{
    if(Front==0)
    {
        cout<<"Queue Underflow"<<endl;
    }
    else
    {
        int Temp;
        Temp=Q[Front];
        cout<<"Deleted "<<Q[Front]<<endl;
        Front++;
    }
}
int Queue::Search()
{
    if(Rear>Size)
    {
        cout<<"Greater Than Size"<<endl;
    }
    else
    {
        cout<<"Enter Search Index=";
        cin>>Rear;
        cout<<Q[Rear]<<" at index "<<Rear<<endl;
    }
}
int Queue::Change()
{
    cout<<"Enter Index to Change=";
    cin>>Rear;
    if(Rear>Size)
    {
        cout<<"Greater Than Size"<<endl;
    }
    else
    {
        cout<<"Enter New Data=";
        cin>>New;
        int temp;
        temp=Q[Rear];
        Q[Rear]=New;
        New=temp;
    }
}
void Queue::display()
{
    for(Rear=Front;Rear<=6;Rear++)
    {
        cout<<"Data "<<Rear<<" ="<<Q[Rear]<<endl;
    }
    cout<<"Front at "<<Front<<" and Rear at "<<Rear<<endl;
}
main()
{
    int s;
    Queue Q1;
    cout<<"1:Enqueue"<<endl;
    cout<<"2:Dequeue"<<endl;
    cout<<"3:Search"<<endl;
    cout<<"4:Change"<<endl;
    cout<<"5:Display"<<endl;
    cout<<"6:Exit"<<endl;
    while(s!=6)
    {
        cout<<"Enter choice:";
        cin>>s;
        switch(s)
        {
            case 1:
                Q1.Enqueue();
                break;
            case 2:
                Q1.Dequeue();
                break;
            case 3:
                Q1.Search();
                break;
            case 4:
                Q1.Change();
                break;
            case 5:
                Q1.display();
                break;
        }
    }
}

File: rust-algos/src/queue.rs
use crate::utils::read_and_parse_data_to_int;

const MAX_SIZE: usize = 10;

pub struct Queue {
    data: [i32; MAX_SIZE],
    front: i32,
    rear: i32,
}

impl Queue {
    fn new() -> Queue {
        Queue {
            data: [0; MAX_SIZE],
            front: -1,
            rear: -1,
        }
    }

    fn enqueue(&mut self, item: i32) {
        let usize_rear: usize = usize::try_from(self.rear).unwrap();

        if usize_rear > MAX_SIZE {
            return println!("Queue limit reached!");
        }

        if self.front == -1 {
            self.front = 0;
        }

        self.rear += 1;
        self.data[usize_rear] = item;

        println!("Inserted!");
    }

    fn dequeue(&mut self) {
        if self.front == self.rear {
            return println!("No data to pop!");
        }

        let usize_top: usize = usize::try_from(self.front).unwrap();

        self.data[usize_top] = 0;
        self.front += 1;

        println!("Removed!");
    }

    fn display(&self) {
        let usize_front: usize = usize::try_from(self.front).unwrap();
        let usize_rear: usize = usize::try_from(self.rear).unwrap();

        println!("Data => {:?}", &self.data[usize_front..usize_rear]);
    }

    pub fn run() {
        let mut stack_data: Queue = Self::new();

        loop {
            println!("\n-------------");
            println!("1: Enqueue");
            println!("2: Dequeue");
            println!("3: Display");
            println!("Other: Exit");
            println!("-------------\n");

            let is_continue: bool =
                read_and_parse_data_to_int("Enter choice: ", |parsed_choice: i32| {
                    match parsed_choice {
                        1 => read_and_parse_data_to_int("Enter item: ", |parsed_item: i32| {
                            stack_data.enqueue(parsed_item);
                            true
                        }),
                        2 => {
                            stack_data.dequeue();
                            true
                        }
                        3 => {
                            stack_data.display();
                            true
                        }
                        _other => {
                            println!("Bye!!\n");
                            false
                        }
                    }
                });

            if !is_continue {
                break;
            }
        }
    }
}
